stl/src/stacktrace.cpp

```diff
@@ -0,0 +1,360 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+// This must be as small as possible, because its contents are
+// injected into the msvcprt.lib and msvcprtd.lib import libraries.
+// Do not include or define anything else here.
+// In particular, basic_string must not be included here.
+
+#include <yvals.h>
+
+#include <cstdio>
+#include <cstdlib>
+
+// clang-format off
+#include <initguid.h> // should be before any header that includes <guiddef.h>
+#include <DbgEng.h>
+#include <Shlwapi.h>
+// clang-format on
+
+#pragma comment(lib, "DbgEng.lib")
+#pragma comment(lib, "Shlwapi.lib")
+
+namespace {
+    // The below function pointer types must be in sync with <stacktrace>
+
+    using _Stacktrace_string_fill_callback = size_t(__stdcall*)(char* _Data, size_t _Size, void* _Context);
+
+    using _Stacktrace_string_fill = size_t(__stdcall*)(
+        size_t _Size, void* _String, void* _Context, _Stacktrace_string_fill_callback _Callback);
+
+    template <class F>
+    size_t string_fill(const _Stacktrace_string_fill callback, const size_t size, void* const str, F f) {
+        return callback(size, str, &f,
+            [](char* s, size_t sz, void* context) -> size_t { return (*static_cast<F*>(context))(s, sz); });
+    }
+
+    // TRANSITION, GH-2285. Use SRWLOCK instead of std::mutex to avoid nontrivial constructor and nontrivial destructor
+    class _NODISCARD srw_lock_guard {
+    public:
+        explicit srw_lock_guard(SRWLOCK& locked_) noexcept : locked(&locked_) {
+            AcquireSRWLockExclusive(locked);
+        }
+
+        ~srw_lock_guard() {
+            ReleaseSRWLockExclusive(locked);
+        }
+
+        srw_lock_guard(const srw_lock_guard&) = delete;
+        srw_lock_guard& operator=(const srw_lock_guard&) = delete;
+
+    private:
+        SRWLOCK* const locked;
+    };
+
+    IDebugClient* debug_client     = nullptr;

```

Comment:
I was also somewhat concerned about how the SRWLOCK needs to protect all accesses to these variables. I agree that enforcing correctness by requiring access to go through the lock guard would be preferable, but I'm somewhat concerned about introducing a whole bunch of function calls into a facility that might be heavily used in non-optimized debug mode (specifically the `operator->()`). Also, introducing a `_Dbg_eng_data_t` and then capturing its address in a `_Lock_guard` would require us to pass that `_Lock_guard` between functions (by reference), since we want to take the lock exactly once per top-level function call, then access the guarded variables repeatedly, even when calling different helper functions.

I think there might be a different "correct by construction" approach:

* Move these variables (including the SRWLOCK) to be static data members of the existing `srw_lock_guard` marked `private`.
* Move the helper functions to be non-static member functions of the guard (like `try_initialize()`, `get_description()`, etc.)
* Then the top-level function calls would construct a guard and call those non-static member functions.

This way we wouldn't pay extra function calls, or have to pass around the address of a data-holder type - the variables are ultimately `static`, we have just one copy of them (hence the need for the SRWLOCK), and the guard object and non-static member functions are ensuring that we acquire/release the SRWLOCK before accessing the static data members.

I'm fine with the code as-is, though. It's also worth noting that because this is within an unnamed namespace within a single `stl/src` file, we can change it arbitrarily in the future without ABI concerns. (The behavior of the top-level `__std_meow` functions must remain unchanged, but they just need to all agree on taking the same lock, so there is no mix-and-match concern if we change the lock - they are within the same file.)

Reply:
Remember that `SRWLOCK` will be `std::mutex` when `mutex` constructor would be `constexpr`, in this case it woild use anusual guard.

I think it is usual to just use locks responible, instead of trying to enforce them.

Reply:
wrapped into structure

Reply:
I've pushed changes to complete this improvement (making the member functions non-static, even though all of the state is static).